### PR TITLE
Added javadoc checks for exported packages.

### LINKED ISSUE #219: Javadoc for methods/variables of exported classes
Currently we have the [JavadocFilterCheck](https://github.com/openhab/static-code-analysis/blob/master/src/main/java/org/openhab/tools/analysis/checkstyle/JavadocFilterCheck.java) which checks **classes** for missing javadoc, but we can add further checks for **methods and variables**.

A good candidate for that may be:
https://github.com/Adobe-Consulting-Services/checkstyle-osgi-checks

>This is a series of Checkstyle checks specific to working in an OSGi environment.
ExportedJavadocMethod
Same as JavadocMethod, but allows missing JavaDoc on non-exported classes.
ExportedJavadocType
Same as JavadocType, but allows missing JavaDoc on non-exported classes.
ExportedJavadocVariable
Same as JavadocVariable, but allows missing JavaDoc on non-exported classes.

It is available on [mvnrepository](https://mvnrepository.com/artifact/com.adobe.acs/checkstyle-osgi-checks) as well.

### PR DESCRIPTION
**Pull request by VelinYordanov**
Closes #219

Signed-off-by: VelinYordanov <velin.iordanov@gmail.com>

### CODE REVIEW

File: custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ExportedJavadocMethodCheck.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import com.puppycrawl.tools.checkstyle.api.Scope;
+import com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocMethodCheck;
+
+/**
+ * Deactivates the JavadocMethodCheck for internal packages and enables it for exported packages.
+ * @author Velin Yordanov - initial contribution
+ *
+ */
+public class ExportedJavadocMethodCheck extends JavadocMethodCheck {
+    @SuppressWarnings("deprecation")
```
Comment (martinvw) [inline custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ExportedJavadocMethodCheck.java]: That sounds bad for a new PR :-)

File: custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/utils/SatCheckUtils.java [hunk 2]
```diff
@@ -98,4 +100,25 @@ public static DetailAST getFirstNodeOfType(DetailAST ast, int type) {
...
+     */
+    public static String getPackageDeclaration(DetailAST root) {
+        DetailAST packageNode = getPackageNode(root);
+        final DetailAST nameAST = packageNode.getLastChild().getPreviousSibling();
+        final FullIdent full = FullIdent.createFullIdent(nameAST);
+        return full.getText();
+    }
+    
+    private static DetailAST getPackageNode(DetailAST node) {
+        if(node.getType() == TokenTypes.PACKAGE_DEF) {
```
Comment (martinvw) [inline custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/utils/SatCheckUtils.java]: Some slight misformatting here

File: custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ExportedJavadocMethodCheck.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.Scope;
+import com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocMethodCheck;
+
+/**
+ * Deactivates the JavadocMethodCheck for internal packages and enables it for exported packages.
+ * @author Velin Yordanov - initial contribution
+ *
+ */
+public class ExportedJavadocMethodCheck extends JavadocMethodCheck {
```
Comment (martinvw) [inline custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ExportedJavadocMethodCheck.java]: Does this only warn about public methods or about all methods?

**Comment by dstivanov** [comment]
@martinvw The PR was reworked due to the usage of some deprecated classes. Can you please take a fresh look again?

File: custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/utils/SatCheckUtils.java (hunk 2 above)
Comment (martinvw) [reply custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/utils/SatCheckUtils.java]: This is still misformatted :-)

File: custom-checks/checkstyle/src/test/java/org/openhab/tools/analysis/checkstyle/test/ExportedJavadocTypeCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import com.puppycrawl.tools.checkstyle.api.Configuration;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtils;
+
+/**
+ * Tests for {@link ExportedJavadocTypeCheck}
+ * @author Velin Yordanov - initial contribution
+ *
+ */
+public class ExportedJavadocTypeCheckTest extends AbstractStaticCheckTest {
+    private final static String LOG_MESSAGE = "javadoc.missing";
```
Comment (martinvw) [inline custom-checks/checkstyle/src/test/java/org/openhab/tools/analysis/checkstyle/test/ExportedJavadocTypeCheckTest.java]: `private static final` please run the sat tool over this PR ;-)

File: custom-checks/checkstyle/src/test/java/org/openhab/tools/analysis/checkstyle/test/ExportedJavadocVariableCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import com.puppycrawl.tools.checkstyle.api.Configuration;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtils;
+
+/**
+ * Tests for {@link ExportedJavadocVariableCheck}
+ * @author Velin Yordanov - initial contribution
+ *
+ */
+public class ExportedJavadocVariableCheckTest extends AbstractStaticCheckTest  {
+    private final static String LOG_MESSAGE = "javadoc.missing";
```
Comment (martinvw) [inline custom-checks/checkstyle/src/test/java/org/openhab/tools/analysis/checkstyle/test/ExportedJavadocVariableCheckTest.java]: See above

File: custom-checks/checkstyle/src/test/resources/checkstyle/exportedJavadocMethodCheckTest/exportedFileWithJavadoc.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.amazondashbutton.handler;
+
+public class AmazonDashButtonHandler extends BaseThingHandler implements PcapNetworkInterfaceListener {
+	public String something;
+	
+	/**
+ * Javadoc
```
Comment (martinvw) [inline custom-checks/checkstyle/src/test/resources/checkstyle/exportedJavadocMethodCheckTest/exportedFileWithJavadoc.java]: Is this intentionally mis formatted like this?

File: custom-checks/checkstyle/src/test/resources/checkstyle/exportedJavadocMethodCheckTest/exportedPackageFile.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.amazondashbutton.handler;
+
+public class AmazonDashButtonHandler extends BaseThingHandler implements PcapNetworkInterfaceListener {
+	public String something;
+    public void handleCommand(ChannelUID channelUID, Command command) {
```
Comment (martinvw) [inline custom-checks/checkstyle/src/test/resources/checkstyle/exportedJavadocMethodCheckTest/exportedPackageFile.java]: Please format all files for this PR

File: sat-plugin/src/main/resources/rulesets/checkstyle/rules.xml [hunk 8]
```diff
@@ -171,6 +171,18 @@
     <module name="IllegalThrows">
       <property name="severity" value="warning" />
     </module>
+    <module name="org.openhab.tools.analysis.checkstyle.ExportedJavadocTypeCheck">
```
Comment (martinvw) [inline sat-plugin/src/main/resources/rulesets/checkstyle/rules.xml]: This XML also looks misformatted

**Pull request closed without merging**